pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -133,183 +122,144 @@ public void init(SegmentGeneratorConfig segmentCreationSpec, SegmentIndexCreatio
+
       if (dictEnabledColumn) {
         // Create dictionary-encoded index
         // Initialize dictionary creator
         // TODO: Dictionary creator holds all unique values on heap. Consider keeping dictionary instead of creator
         //       which uses off-heap memory.
-        SegmentDictionaryCreator dictionaryCreator =
-            new SegmentDictionaryCreator(fieldSpec, _indexDir, columnIndexCreationInfo.isUseVarLengthDictionary());
-        _dictionaryCreatorMap.put(columnName, dictionaryCreator);
-        // Create dictionary
+
+        DictionaryIndexType dictIdx = DictionaryIndexType.INSTANCE;
+        // Index conf should be present if dictEnabledColumn is true. In case it doesn't, getConfig will throw an
+        // exception
+        DictionaryIndexConfig dictConfig = config.getConfig(dictIdx);
+        if (!dictConfig.isEnabled()) {
+          throw new IllegalArgumentException("Dictionary index should be enabled");
+        }
+        SegmentDictionaryCreator creator = dictIdx.createIndexCreator(context, dictConfig);
+
         try {
-          dictionaryCreator.build(columnIndexCreationInfo.getSortedUniqueElementsArray());
+          creator.build(context.getSortedUniqueElementsArray());
         } catch (Exception e) {
           LOGGER.error("Error building dictionary for field: {}, cardinality: {}, number of bytes per entry: {}",
-              fieldSpec.getName(), columnIndexCreationInfo.getDistinctValueCount(),
-              dictionaryCreator.getNumBytesPerEntry());
+              context.getFieldSpec().getName(), context.getCardinality(), creator.getNumBytesPerEntry());
           throw e;
         }
-      }
 
-      if (bloomFilterColumns.contains(columnName)) {
-        if (indexingConfig.getBloomFilterConfigs() != null
-            && indexingConfig.getBloomFilterConfigs().containsKey(columnName)) {
-          _bloomFilterCreatorMap.put(columnName, _indexCreatorProvider.newBloomFilterCreator(
-              context.forBloomFilter(indexingConfig.getBloomFilterConfigs().get(columnName))));
-        } else {
-          _bloomFilterCreatorMap.put(columnName, _indexCreatorProvider.newBloomFilterCreator(
-              context.forBloomFilter(new BloomFilterConfig(BloomFilterConfig.DEFAULT_FPP, 0, false))));
-        }
+        _dictionaryCreatorMap.put(columnName, creator);
       }
 
-      if (!columnIndexCreationInfo.isSorted() && rangeIndexColumns.contains(columnName)) {
-        _rangeIndexFilterCreatorMap.put(columnName,
-            _indexCreatorProvider.newRangeIndexCreator(context.forRangeIndex(rangeIndexVersion)));
+      Map<IndexType<?, ?, ?>, IndexCreator> creatorsByIndex =
+          Maps.newHashMapWithExpectedSize(IndexService.getInstance().getAllIndexes().size());
+      for (IndexType<?, ?, ?> index : IndexService.getInstance().getAllIndexes()) {
+        if (skipIndexType(index)) {
```

Comment:
extend skipIndexType() to include checks: config.isEnabled and index.shouldBeCreated too? 
 
nit: call it tryCreaetIndexCreator for a bit more clarity

Reply:
> extend skipIndexType() to include checks: config.isEnabled and index.shouldBeCreated too?

I don't like that because the idea of `skipIndexType` was to only contain the indexes whose lifecycle is different than the others in order to make this difference more explicit. I guess I didn't achieve that target given that the name of the method is too generic.

What I did was to rename the method to `hasSpecialLifecycle`. I also removed the redundant config.isEnabled check, as it was already inclued in tryCreateCreator. I also renamed that method as `tryCreateIndexCreator` as you suggested.